✨ Feat: ELBv2/TGs - Add health check customization

mtulio opened the pull request:
**What type of PR is this?**

/kind feature
/kind api-change

**What this PR does / why we need it**:

<!-- Enter a description of the change and why this change is needed -->

This change exposes the health check configuration for listeners of both load balancers, primary and secondary. When the target group is created for the API, the user will be able to customize only the health check probes, according to the limit ranges [allowed by AWS](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/target-group-health-checks.html#health-check-settings).

It will allow providers to customize the API and additional listeners' target health checks to ensure existing implementations.

A significant improvement will be in the additional listeners which is currently set with basic health checks following the same protocol of the listener (TCP). Exposing this value will allow customized health check.

The example below shows how we are using to customize the API target group (from the default provided by CAPA), alongside setting custom health check parameters for additional listeners, like overriding the protocol to `HTTPS`, check path `/healthz`, setting custom probe timers:

```go
awsCluster := &capa.AWSCluster{
  ObjectMeta: metav1.ObjectMeta{
	Name:      clusterID.InfraID,
	Namespace: capiutils.Namespace,
  },
  Spec: capa.AWSClusterSpec{
	ControlPlaneLoadBalancer: &capa.AWSLoadBalancerSpec{
		Name:                ptr.To(clusterID.InfraID + "-int"),
		LoadBalancerType:    capa.LoadBalancerTypeNLB,
		Scheme:              &capa.ELBSchemeInternal,
                HealthCheckProtocol: &capa.ELBProtocolHTTPS,
		HealthCheck: &capa.TargetGroupHealthCheck{
			IntervalSeconds:         ptr.To(int64(10)),
			TimeoutSeconds:          ptr.To(int64(10)),
			ThresholdCount:          ptr.To(int64(2)),
			UnhealthyThresholdCount: ptr.To(int64(2)),
		},
		AdditionalListeners: []capa.AdditionalListenerSpec{
			{
				Port:     22623,
				Protocol: capa.ELBProtocolTCP,
				HealthCheck: &capa.TargetGroupHealthCheck{
					Protocol:                ptr.To("HTTPS"),
					Path:                    ptr.To("/healthz"),
					IntervalSeconds:         ptr.To(int64(10)),
					TimeoutSeconds:          ptr.To(int64(10)),
					ThresholdCount:          ptr.To(int64(2)),
					UnhealthyThresholdCount: ptr.To(int64(2)),
				},
			},
		},
	},
	SecondaryControlPlaneLoadBalancer: &capa.AWSLoadBalancerSpec{
		Name:                ptr.To(clusterID.InfraID + "-ext"),
		LoadBalancerType:    capa.LoadBalancerTypeNLB,
		Scheme:              &capa.ELBSchemeInternetFacing,
                HealthCheckProtocol: &capa.ELBProtocolHTTPS,
		HealthCheck: &capa.TargetGroupHealthCheck{
			IntervalSeconds:         ptr.To(int64(10)),
			TimeoutSeconds:          ptr.To(int64(10)),
			ThresholdCount:          ptr.To(int64(2)),
			UnhealthyThresholdCount: ptr.To(int64(2)),
		},
        },
  },
}
```

This PR is creating dedicated the health check structures, a more restricted one to API and field-free for additional listeners, protecting to expose the API target group for both load balancers when changing standardized fields like protocol, port and path..

**Which issue(s) this PR fixes**:

Fixes #4884


**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [x] includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)
- [x] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Exposing the health check attributes for the target group for the control plane load balancers, allowing customized health checks for API or additional listeners.
```

mtulio commented:
fwiw @vincepri @patrickdillon @r4f4

mtulio commented:
The provision of API and additional listeners are working as expected, I am investigating unit failures for fuzzy test `TestFuzzyConversion`[2]

[1] e2e deployment w/ custom target health check

![Screenshot from 2024-03-05 23-34-49](https://github.com/kubernetes-sigs/cluster-api-provider-aws/assets/3216894/af99e9e2-6aef-4f3b-a6b5-d4405e6080e5)
![Screenshot from 2024-03-05 23-35-46](https://github.com/kubernetes-sigs/cluster-api-provider-aws/assets/3216894/ba9e8bef-d2c1-4a6d-b8b8-3a2a62e41f88)

[2] fuzzy test failures


```
--- FAIL: TestFuzzyConversion (16.75s)
    --- FAIL: TestFuzzyConversion/for_AWSCluster (3.69s)
        --- FAIL: TestFuzzyConversion/for_AWSCluster/hub-spoke-hub (0.00s)
            conversion.go:241: 
                  &v1beta2.AWSCluster{
...
                - 		Annotations:                nil,
                + 		Annotations:                map[string]string{},
...
                - 			AdditionalControlPlaneIngressRules: []v1beta2.IngressRule{},
                + 			AdditionalControlPlaneIngressRules: nil,
...
                - 			HealthCheck:              &v1beta2.TargetGroupHealthCheck{Path: &"ĆĠ成", Port: &"缆zp%!ʛ(MISSING)(", TimeoutSeconds: &5940725222272119504},
                + 			HealthCheck:              nil,
...
                - 				AvailabilityZones: []string{},
                + 				AvailabilityZones: nil,
...
FAIL
FAIL	sigs.k8s.io/cluster-api-provider-aws/v2/api/v1beta1	16.761s

```

nrb commented on api/v1beta2/network_types.go:
```diff
@@ -94,12 +104,57 @@ var (
+	// Defaults to TCP.
+	// +kubebuilder:default=TCP
+	// +kubebuilder:validation:Enum=TCP;HTTP;HTTPS
+	// +optional
```
How does defaulting the field will interact with existing entries that don't have a value? I'm not super familiar with controller-gen, but it [looks like](https://github.com/kubernetes-sigs/controller-tools/blob/020c71865aa9bf42a25e0cf3cf0dedbf435694dd/pkg/crd/markers/validation.go#L485) the default value is written to the Schema, which would then be applied to objects where the field is empty.

nrb replied:
Yes, that would resolve my concern.

faiq commented on api/v1beta2/network_types.go:
```diff
@@ -94,12 +104,57 @@ var (
+	// Defaults to "/readyz".
+	// +kubebuilder:default="/readyz"
+	// +optional
+	Path *string `json:"path,omitempty"`
```
I really don't think exposing this is a great idea, and in fact could be a potential footgun for users. 

We would need to add additional logic, either in the form of a [cel statement](https://kubernetes.io/docs/reference/using-api/cel/) or a webhook to ensure that users are not able to put together an invalid combination ie: using `protocol: ssl` and a path. 

Furthermore, for path there's only one correct answer for a health check when it comes to the kube-api-server which is `/readyz`. If a user puts in another path this can result in an error.

Overall, I don't think all of this needs to be exposed to the user and can open a lot of edge cases.

mtulio replied:
@faiq +1 for limiting and protecting the TG health check for the Kube API.

The initial idea of those changes are to keep compatibility with existing deployment using HTTPS health check when migrating to CAPA. The fix proposed #4846 cover the TG for API, this PR will intent to cover the needed to customize the health check path for the additional listeners[1] - which is currently default to the target protocol.

We also need to customize other health check attributes like interval, threshold, etc. Do you think we can expose those[2] health check attributes, limiting the API path to always use `/readyz`, allowing customization for other parameters?

[1] https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4849/files#diff-a565eb62d3c745e0d2f5137e2a9f8fbffc335675576635f62a30054a0d2375a9L249

[2] https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4849/files#diff-1c519200705fb576b64e7fb271f7f683a2e6efadd54abd2a70b243e5069490e9R127-R157

faiq replied:
yeah,  maybe port and path are things that do not need to be configured by the user because they are fixed. 

Regarding the other fields we should take the validation from [AWS directly](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/target-group-health-checks.html#health-check-settings) if possible

mtulio replied:
> Regarding the other fields we should take the validation from [AWS directly](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/target-group-health-checks.html#health-check-settings) if possible

@faiq I wonder if we take the validation from AWS, and not the [constants `DefaultAPIServerHealth*`](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/cbf531711d117b823c587622ba0e64b94d0fd314/api/v1beta2/network_types.go#L31-L39) may cause API changes, thougts?

faiq replied:
I don't think it would make API changes?  These would be `kubebuilder:validation` comments to verify ranges that AWS uses

faiq replied:
Can we not field expose this to the user?

mtulio replied:
@faiq considering the `TargetGroupHealthCheck` is used for all listeners, KAPI and additional, and additional can be any user-defined service which would be allowed to provide customized Path, instead of default "Protocol + Port" inherited from Listener: I am not sure if we can provide a kube builder rule to restrict that level, do you know?

Another option would be duplicate the `TargetGroupHealthCheck` to additional listeners exposing almost all fields for those custom listeners, and restrict for the Kube API.

WDYT? Do you have any other suggestion?

mtulio replied:
> I am not sure if we can provide a kube builder rule to restrict that level, do you know?

You've mentioned about creating a ceil validation, it seems to be applied to this question too. I will research about this.

mtulio replied:
@faiq WRT this thread: I changed the approach of exposing the API target group attributes by limiting the probe fields based in the AWS range  values.

Instead of exposing the entire HealthCheck structure, I am creating the Health check receipt for health check for API and one for additional listeners, where API is more restricted. Those receipts have kubebuilder validation for ranges allowed by AWS. I am also not setting the defaults to keep compatibility.

cc/fwiw @vincepri @nrb

faiq replied:
i think all this looks good. thanks for considering my feedback!

mtulio commented:
PR comments addressed.
The issue was already created to formalize the "feature request" flow :)

https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4884

cc @nrb @vincepri @faiq

mtulio commented:
Hey @faiq - any tips to fix the [Fuzzy][fuzzy] unit test failing?
I am not sure if it is related with [this conversion][conversion], but honestly I didn't understand enough as it is pointing to v1beta1 which I am not changing (should I need it?)

[fuzzy]: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4849/pull-cluster-api-provider-aws-test/1772863309742084096#1:build-log.txt%3A369-372
[conversion]: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4849/pull-cluster-api-provider-aws-verify/1772863309955993600#1:build-log.txt%3A755

faiq commented:
I think some code needs to be added to the conversion. I'll see what I can find.

mtulio commented:
Converting to draft to make the `pull-cluster-api-provider-aws-verify` green without re-running all others consuming infra time.

mtulio commented:
/test pull-cluster-api-provider-aws-verify

mtulio commented:
> /test pull-cluster-api-provider-aws-verify

Passing. Converting to regular PR

faiq reviewed (commented):
/lgtm

mtulio commented:
Hey Vince, PTAL?
/assign @vincepri

mtulio commented:
/test pull-cluster-api-provider-aws-e2e

mtulio commented:
> /lgtm

@faiq changes by mistake cleaned the `lgtm` label, apologies. It was already reverted. Would you mind stamping it again?

mtulio commented:
/test pull-cluster-api-provider-aws-e2e

mtulio commented:
Found one case when `UnhealthThresholdCount` attribute isn't updated. Fixed and added a unit case to get it.

/assign @nrb @nrb

mtulio commented:
Changes from last LGTM: in the PR #4846 we standardized the API target group's health check attributes, the [latest change here defines the default attributes explicitly](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4849/files#diff-ee38bda5efe2606e7e7746a556c49a2d53d14eca60c029b8efebea657b8718f1R409-R418), allowing the customization (this feature) when the spec is defined.

mtulio commented:
> Found one case when `UnhealthThresholdCount` attribute isn't updated. Fixed and added a unit case to get it.
> 
> /assign @nrb @nrb

Changes from last LGTM: in the PR #4846 we standardized the API target group's health check attributes, the [latest change here defines the default attributes explicitly](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4849/files#diff-ee38bda5efe2606e7e7746a556c49a2d53d14eca60c029b8efebea657b8718f1R409-R418), allowing the customization (this feature) when the spec is defined.

mtulio commented:
/test pull-cluster-api-provider-aws-e2e

rvanderp3 commented:
/lgtm

nrb commented:
/lgtm as well

mtulio commented:
@richardcase would you mind taking a look im this proposal, please?

richardcase commented:
/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-eks

richardcase commented:
Thanks for this @mtulio . 

Once both the e2e passes we can unhold

/hold

From my side this looks good, so:

/approve

richardcase commented:
Timed out, retrying in case its a flake:

/test pull-cluster-api-provider-aws-e2e

nrb commented:
Yet another time out with GPU instances...

/test pull-cluster-api-provider-aws-e2e

faiq commented:
@nrb is there a bug/issue for the GPU tests?

nrb commented:
> @nrb is there a bug/issue for the GPU tests?

Not specifically yet, just #4937 and https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4783#issuecomment-2066419854

nrb commented:
/unhold

r4f4 commented:
@mtulio seems like the same problem here: needs a rebase on top of master.

mtulio commented:
> @mtulio: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-cluster-api-provider-aws-verify	[97489d2](https://github.com/kubernetes-sigs/cluster-api-provider-aws/commit/97489d20d97f65eb603fc7ed340a0e06ecb55378)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4849/pull-cluster-api-provider-aws-verify/1781384409866309632)	unknown	`/test pull-cluster-api-provider-aws-verify`

Rebased to [fix codegen issue](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4882#issuecomment-2066893277)

mtulio commented:
@nrb passing now, after rebased. Would you mind to res-stamp lgtm :)

nrb commented:
/lgtm
